### ISSUE TITLE
Cancel popstate navigations on certain conditions

### ISSUE THREAD

**Issue by skelesp** [angular/angular#40371]
# 🚀 feature request

### Relevant Package
@angular/router
@ng-bootstrap/ng-bootstrap

### Description
I have modals on almost every route in my app. When a user clicks on browser back or forward while a modal is open, I don't want him to navigate to the previous page. What I want is to close the modal and stay on the current page...
This is my solutinn so far... But this messes up the browser navigation history ...

```typescript
constructor (
    private router: Router,
    private modalService: NgbModal
  ) { }

  ngOnInit(): void {
    this.routerEventSubscription = this.router.events.pipe(
      filter( event => event instanceof NavigationStart ),
      pairwise(),
      tap(events => console.log(events))
    ).subscribe( ([previousEvent, currentEvent]: [NavigationStart, NavigationStart]) => {
      if (this.modalService.hasOpenModals()) {
        this.modalService.dismissAll();
        if (currentEvent.navigationTrigger === 'popstate') {
          this.router.navigateByUrl(previousEvent.url, { replaceUrl: true });
          console.warn("User clicked back or forward in browser with modal open, so modal is closed and return to page from which modal was opened.");
        }
      }
    })
  }
  
  ngOnDestroy () {
    this.routerEventSubscription.unsubscribe();
  }
```

### Describe the solution you'd like
Some way to subscribe to router events in my app.component and cancel the current navigation on certain conditions without having to add a route guard to every route in my app and without messing up the history.

In AngularJS, I managed to achieve this with following code in app.run() , so I would like to be able to do the same in Angular...

```javascript
 $rootScope.$on('$locationChangeStart', function ($event, newUrl, oldUrl) {
        var openedModal = $uibModalStack.getTop();
        if (openedModal) {
	        if (!!$event.preventDefault) {
		        $event.preventDefault();
	        }
	        if (!!$event.stopPropagation) {
		        $event.stopPropagation();
	        }
	        $uibModalStack.dismiss(openedModal.key);
        }
});
```

### Describe alternatives you've considered
Putting a route guard on every route in my app. But this is not a scalable solution...

**Comment by atscott** [comment]
I believe what you want to do is to actually change the url when the popup is opened. This can be done in several ways, one of which would be named outlets as described here: https://angular.io/guide/router-tutorial-toh#displaying-multiple-routes-in-named-outlets. That way, when you're clicking the back or forward button, you're navigating through the history the way you'd expect and you can use a guard to prevent the navigation (unfortunately this is actually broken at the moment #13586).

**Comment by skelesp** [comment]
Indeed, named outlets might be a way to implement this. But that's something different than what I was able to do in angularJS.

Is there a way to do something like I did in AngularJS? (See code above)

**Comment by atscott** [comment]
@skelesp What you have seems, more or less, reasonable (though I'm not sure about the `replaceUrl: true`). Triggering a new navigation will cancel the in-flight one. I can see why this would mess up the history though because the in-flight navigation was triggered to synchronize the internal router state with the browser URL. It's possible that your current solution would be fixed when #13586 is resolved; the cancelled navigation should result in restoring the browser history better by triggering a forward navigation in the browser history if the "back" was cancelled.
